### ISSUE TITLE
no bugs being reported

### ISSUE THREAD

**Issue by hutcj** [Riverside-Software/sonar-openedge#761]
Hi all - new here. I've gone through the setup instructions a number of times but needed some help. Just spun up a SonarQube server (used the docker image, still using the embedded database for now), installed the OpenEdge plugin, and installed sonar-scanner on the same machine.

After cloning the source code (& its dependencies and copies of the DB schema) to the SonarQube server, I created a sonar-project.properties file in the repository and set values for the following settings: sonar.projectKey, sonar.sources, sonar.sourceEncoding, sonar.projectName, sonar.projectVersion, sonar.projectDescription, sonar.oe.propath, sonar.oe.databases, sonar.oe.aliases, and sonar.oe.proparse.debug.

The problem is that even after several runs of sonar-scanner and waiting for the analysis to complete on the SonarQube server (from the web GUI), there are zero bugs, code smells, etc. I even purposely put some invalid code in but that was not caught. SonarQube appears to see the code base properly - I can browse the folders and it shows 782k lines of code, 28.3% duplication, 18k duplicated blocks; it just won't show any bugs. Are the rules not being applied somehow? (I only installed the free rule set.)

Does OpenEdge have to be installed on the SonarQube server? I see references to DLC and standard Progress includes for some PROPATH stuff.

We have not started using ANT/PCT for our build process yet, but the docs seemed to suggest that some rules should still work properly even if we don't use those tools yet. 

Thanks!

**Comment by gquerret** [comment]
The sonar-openedge plugin comes by default with only a very limited number of rules (warnings raised by the compiler, code duplication and failing unit tests). But the first one (compiler warnings) needs the output of PCT (as warnings are expected in a specific format).
So with the free ruleset, you can write your own rules (a sample plugin is available [here](https://github.com/Riverside-Software/rules-plugin-template) ) or you can ask for an evaluation license of the commercial set of rules (instructions are available in the Wiki).

And if you introduced invalid code, you should see this kind of output (near the end of the scanner report):
```
INFO: 31 files proparse'd, 22 XML files, 22 listing files, 0 failure(s), 800 NCLOCs
```
The number of failures indicates the files where the parser failed. A high ratio usually means that some properties are incorrectly set (propath, db connections, aliases)

**Comment by hutcj** [comment]
Thanks, @gquerret 
I see - so until we start using PCT, this is working as intended? We do not have any unit testing yet. I will send a evaluation license request later.

Does Progress have to be installed on the SonarQube server? Should I use PCT to compile on our development server that has Progress installed on it and then transfer the output to the SonarQube server for analytics? I'm having trouble understanding where this fits in.

I ran 2 additional tests - one with intentional errors and one with the code base as-is (code base has roughly 4-5k procedures).

With intentional errors, I saw this in the scanner output:
```
...
INFO: 4466 files proparse'd, 0 XML files, 0 listing files, 0 failure(s), 782288 NCLOCs
...
INFO: 1 files to be analyzed
INFO: 0/1 files analyzed
WARN: Missing blame information for the following files:
WARN: <name of file that I edited>
...
```

With the normal run, I saw this
```
...
INFO: 4466 files proparse'd, 0 XML files, 0 listing files, 0 failure(s), 782285 NCLOCs
...
INFO: 4467 files to be analyzed
...
INFO: 4467/4467 files analyzed
...
```

Both of the runs also showed errors like this:
```
ERROR: Syntax error -- <procedure-file-name>.p -- <include-file-name>.i:15:48 -- no viable alternative at input 'cursor-down of  do' -- Recover
ERROR: Syntax error <unknown>:44:10 mismatched input 'today' expecting {'?', QUOTED_STRING}
ERROR: Syntax error <unknown>:58:10 mismatched input 'NOW' expecting {'?', QUOTED_STRING}
```

There was also a Java stack trace dumped after this error:
```
ERROR: Unable to lex file '<include-file-name>.i'
org.prorefactor.core.ProparseRuntimeException: Proparse error. Infinite loop caught by preprocessor.
```

**Comment by gquerret** [comment]
> I see - so until we start using PCT, this is working as intended? We do not have any unit testing yet. I will send an evaluation license request later.

Yes

> Does Progress have to be installed on the SonarQube server?

You don't have to. But your source may need Progress include files (mostly adecomm/ or web/ and web2), so this is why there's the `sonar.oe.dlc` property.

> I'm having trouble understanding where this fits in.

Using Jenkins (but this applies to any other CI server), the SonarQube analysis follows the build, so it's executed on the same server. Communication with the SQ server is done over HTTP, so the SQ server can be installed anywhere.

> ERROR: Syntax error -- <procedure-file-name>.p -- <include-file-name>.i:15:48 -- no viable alternative at input 'cursor-down of  do' -- Recover
ERROR: Syntax error <unknown>:44:10 mismatched input 'today' expecting {'?', QUOTED_STRING}
ERROR: Syntax error <unknown>:58:10 mismatched input 'NOW' expecting {'?', QUOTED_STRING}

By default, the parser tries to inject one keyword or remove one keyword to recover from parse failures. So in this case, the parser was able to recover, and no error was raised.

> ERROR: Unable to lex file '<include-file-name>.i'
org.prorefactor.core.ProparseRuntimeException: Proparse error. Infinite loop caught by preprocessor

Would you mind sharing this include file ? You can do that privately at gquerret@riverside-software.fr
